lib/Optimizer/Transforms/LiftArrayAlloc.cpp

```diff
@@ -321,6 +325,30 @@ class ComplexCreatePattern : public OpRewritePattern<complex::CreateOp> {
   }
 };
 
+class CustomUnitaryPattern
+    : public OpRewritePattern<quake::CustomUnitarySymbolOp> {
+
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(quake::CustomUnitarySymbolOp customOp,
+                                PatternRewriter &rewriter) const override {
+    auto sref = customOp.getGenerator();
+    StringRef generatorName = sref.getRootReference();
+    auto parentModule = customOp->getParentOfType<ModuleOp>();
+
+    auto ccGlobalOp = parentModule.lookupSymbol<cudaq::cc::GlobalOp>(
+        generatorName.str() + ReadOnlySuffix);
```

Comment:
This lookup isn't going to be general enough because when we lift array values, we don't just look for the special unitary functions. And we cannot restrict it that way.

This is a general pass and shouldn't be a one-off just for unitary matrices. So we need a new, separate pass to rewrite the custom unitaries.

I think the algorithm here has to be:
  - look up the generator function, which should be the symbol before we rewrite anything.
  - scan the function's body for the address_of operation and the symbol it is using.
  - use that symbol to replace the current op with a new custom op. (yes, we want to replace it, otherwise we can have mysterious crashes in the transformer)

Reply:
Addressed in d16052be0a40ca2b4f8ce19b8b8b9df53cd8a537